File: 240426/등장하지 않는 문자열의 길이/length-of-string-that-does-not-appear.py
N = int(input())
n_str = input()

pattern_cnts = {}
for s_len in range(1, N):
    for chr_idx in range(N - s_len + 1):
        if n_str[chr_idx:chr_idx + s_len] not in pattern_cnts:
            pattern_cnts[n_str[chr_idx:chr_idx + s_len]] = 1
        else:
            pattern_cnts[n_str[chr_idx:chr_idx + s_len]] += 1

patterns = []
max_length = max(map(len, pattern_cnts.keys()))

for length in range(1, max_length + 1):
    tmp = []
    for pattern, cnts in pattern_cnts.items():
        if len(pattern) == length:
            tmp.append((pattern, cnts))

    if tmp:
        patterns.append(tmp)

min_length = 0
patterns_sort = sorted(patterns, key = lambda x: x[0][1], reverse=True)
for pattern_lst in patterns_sort:
    patterns, pattern_cnts = list(zip(*pattern_lst))[0], list(zip(*pattern_lst))[1]
    if 2 in pattern_cnts:
        continue
    else:
        min_length = len(patterns[0])
        break

print(min_length)

File: 260321/등장하지 않는 문자열의 길이/length-of-string-that-does-not-appear.cpp
#include <bits/stdc++.h>
using namespace std;

int main() 
{
    int N;
    cin >> N;

    string n_str;
    cin >> n_str;

    map<string, int> pattern_cnts;

    for (int s_len = 1; s_len < N; s_len++) 
    {
        for (int chr_idx = 0; chr_idx <= N - s_len; chr_idx++) 
        {
            string sub = n_str.substr(chr_idx, s_len);
            pattern_cnts[sub]++;
        }
    }
    // Please write your code here.
    vector<vector<pair<string, int>>> patterns;
    int max_length = 0;
    for (const auto& p : pattern_cnts) 
    {
        max_length = max(max_length, (int)p.first.size());
    }

    for (int length = 1; length <= max_length; length++) 
    {
        vector<pair<string, int>> tmp;
        for (const auto& p : pattern_cnts) 
        {
            if ((int)p.first.size() == length) 
            {
                tmp.push_back({p.first, p.second});
            }
        }
        if (!tmp.empty()) 
        {
            patterns.push_back(tmp);
        }
    }

    sort(patterns.begin(), patterns.end(),
         [](const vector<pair<string, int>>& a, const vector<pair<string, int>>& b) 
         {
             return a[0].second > b[0].second;
         });

    int min_length = 0;

    for (const auto& pattern_lst : patterns) 
    {
        bool all_less_than_2 = true;
        for (const auto& p : pattern_lst) 
        {
            if (p.second >= 2) 
            {
                all_less_than_2 = false;
                break;
            }
        }

        if (all_less_than_2) 
        {
            min_length = (int)pattern_lst[0].first.size();
            break;
        }
    }

    cout << min_length << '\n';
    return 0;
}